src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp

```diff
@@ -0,0 +1,167 @@
+    const double r_old =
+        sqrt(square(x_old[0]) + square(x_old[1]) + square(x_old[2]));
+    const double theta_old = acos(x_old[2] / r_old);
+    const double phi_old = atan2(x_old[1], x_old[0]);
+
+    // Evaluate the radius of the surface on the old strahlkorper
+    // at theta_old, phi_old.
+    const double r_surf_old = strahlkorper->radius(theta_old, phi_old);
+
+    // If r_surf_old is r_old, then 'r_new' is on the surface.
+    return r_surf_old - r_old;
+  };
+
+  // We try to bracket the root between r_min and r_max.
+  // But r_min and r_max are only approximate, and there may be grid points
+  // with radii outside that range. So we pad by 10% to be safe.
+  const double padding = 0.10;
+
+  const auto radius_at_each_angle = RootFinder::toms748(
+      radius_function,
+      make_with_value<DataVector>(get<0>(r_hat), r_min * (1.0 - padding)),
+      make_with_value<DataVector>(get<0>(r_hat), r_max * (1.0 + padding)),
+      std::numeric_limits<double>::epsilon() * (r_min + r_max),
+      2.0 * std::numeric_limits<double>::epsilon());
+
+  // Now reset the radius and center of the new strahlkorper.
+  *strahlkorper =
+      Strahlkorper<Frame>(strahlkorper->l_max(), strahlkorper->m_max(),
+                          radius_at_each_angle, new_center);
+}
+}  // namespace change_expansion_center_of_strahlkorper_detail
+
+template <typename Frame>
+void change_expansion_center_of_strahlkorper(
+    gsl::not_null<Strahlkorper<Frame>*> strahlkorper,
+    const std::array<double, 3>& new_center) noexcept {
+  change_expansion_center_of_strahlkorper_detail::change_expansion_center(
+      strahlkorper, new_center,
+      change_expansion_center_of_strahlkorper_detail::get_rhat(*strahlkorper));
+}
+
+template <typename Frame>
+void change_expansion_center_of_strahlkorper_to_physical(
+    gsl::not_null<Strahlkorper<Frame>*> strahlkorper) noexcept {
+  const auto r_hat =
+      change_expansion_center_of_strahlkorper_detail::get_rhat(*strahlkorper);
+
+  // Zeroth iteration.
+  change_expansion_center_of_strahlkorper_detail::change_expansion_center(
+      strahlkorper, strahlkorper->physical_center(), r_hat);
+
+  // In the random number tests, it never needed more than 7
+  // iterations to converge to relative error of roundoff.  Allow 14
+  // iterations to be safe.
+  const size_t maxiter = 14;
+  for (size_t iter = 0; iter < maxiter; ++iter) {
+    const auto phys_center = strahlkorper->physical_center();
+    const auto center = strahlkorper->center();
+    const auto average_radius = strahlkorper->average_radius();
+    const double relative_error = sqrt(square(center[0] - phys_center[0]) +
+                                       square(center[1] - phys_center[1]) +
+                                       square(center[2] - phys_center[2])) /
+                                  average_radius;
+    if (relative_error <= std::numeric_limits<double>::epsilon()) {
```

Comment:
This is very strict. How sure are you that something like `2.0 * epsilon` wouldn't be better?

Reply:
Yes it is strict.  I'm happy to change it to 2.0*epsilon if you want but I have no way of justifying why it should be 2.0 versus 3.0 or 8.0.   So far just epsilon did not cause any of the random number tests to fail. Convergence is very rapid for the tests that I have done, 7 iterations at most.

Reply:
Okay, my worry is just that it might be much easier to achieve 2 * eps simply because you are trying to get all the bits to match. I agree there isn't a reason for why a specific value times eps, but I would be more comfortable not having us trying to match something to bitwise machine precision. Having said that, it's not something I would block the PR for, just something I wanted to generate some short discussion over :)

Reply:
Ok, changed to 2.0*epsilon.  It will probably save an iteration in some cases.